src/pgrn-custom-scan.c

```diff
@@ -181,19 +183,59 @@ PGrnBeginCustomScan(CustomScanState *customScanState,
 					EState *estate,
 					int eflags)
 {
-	grn_obj *sourcesTable;
+	PGrnScanState *state = (PGrnScanState *) customScanState;
+	TupleDesc tupdesc =
+		RelationGetDescr(customScanState->ss.ss_currentRelation);
 	Relation index = PGrnChooseIndex(customScanState->ss.ss_currentRelation);
+	grn_obj *sourcesTable;
+
 	if (!index)
 		return;
 	sourcesTable = PGrnLookupSourcesTable(index, ERROR);
-	elog(LOG, "DEBUG: custom-scan: %s", PGrnInspect(sourcesTable));
 	RelationClose(index);
+
+	GRN_PTR_INIT(&(state->columns), GRN_OBJ_VECTOR, GRN_ID_NIL);
+	for (unsigned int i = 0; i < tupdesc->natts; i++)
+	{
+		Form_pg_attribute attr = TupleDescAttr(tupdesc, i);
+		grn_obj *column =
+			PGrnLookupColumn(sourcesTable, NameStr(attr->attname), ERROR);
+		GRN_PTR_PUT(ctx, &(state->columns), column);
+	}
+	state->tableCursor = grn_table_cursor_open(
+		ctx, sourcesTable, NULL, 0, NULL, 0, 0, -1, GRN_CURSOR_ASCENDING);
 }
 
 static TupleTableSlot *
-PGrnExecCustomScan(CustomScanState *node)
+PGrnExecCustomScan(CustomScanState *customScanState)
 {
-	return NULL;
+	PGrnScanState *state = (PGrnScanState *) customScanState;
+	grn_id id;
+
+	if (!state->tableCursor)
+		return NULL;
+
+	id = grn_table_cursor_next(ctx, state->tableCursor);
+	if (!id)
+		return NULL;
+
+	{
+		TupleTableSlot *slot = customScanState->ss.ps.ps_ResultTupleSlot;
+		grn_obj value;
+
+		ExecClearTuple(slot);
+		GRN_VOID_INIT(&value);
+		for (unsigned int i = 0; i < slot->tts_tupleDescriptor->natts; i++)
+		{
+			Form_pg_attribute attr = &(slot->tts_tupleDescriptor->attrs[i]);
+			grn_obj *column = GRN_PTR_VALUE_AT(&(state->columns), i);
+			grn_obj_get_value(ctx, column, id, &value);
+			slot->tts_values[i] = PGrnConvertToDatum(&value, attr->atttypid);
+			slot->tts_isnull[i] = false;
```

Comment:
TODO: We need to reject our custom scan if there is at least one nullable column like we did for index only scan: https://github.com/pgroonga/pgroonga/blob/cac1dbb0fc4aa515595760ad2a0aa3e060594cdd/src/pgroonga.c#L8519